### PR TITLE
Store domain and its value in summary

### PR DESCRIPTION
**Pull request by FrNecas**
This is a preparation for implementation of GOTO code instrumentation. It will be useful to have access to the whole domain to be able to easily instrument GOTO code with invariants. It was necessary to slightly tweak how summary_db works with summaries (previously, summaries were copied when obtaining them from summary_db, references are returned after the changes).

### CODE REVIEW

**Review by viktormalik — CHANGES_REQUESTED** [review]

File: src/2ls/graphml_witness_ext.cpp
Comment (FrNecas) [reply src/2ls/graphml_witness_ext.cpp:37]: That's what I wanted to do initially but it didn't work. Assigning to a reference performs a copy which is in conflict with the unique_ptr to domain in summary. That's why I decided to put the reference creation and `cfg()` call into the `if-else`. Due to this change, it was necessary to shift `ssa_unwinder` and `cfg` definitions above the if. However, I've noticed that I made some unnecessary changes while doing so, I will remove them and fix the commit.

**Review by peterschrammel — COMMENTED** [review]
@viktormalik,I guess this should wait until you've merged your SV-COMP fixes?

File: src/solver/summary.h
Comment (viktormalik) [reply src/solver/summary.h:58]: Because the domain stores the variables and the the value stores the abstract values for the variables. We need both for instrumentation of GOTO programs with invariants.

**Comment by viktormalik** [comment]
> @viktormalik,I guess this should wait until you've merged your SV-COMP fixes?

There shouldn't be any conflict but I agree, let's merge the SV-COMP fixes first (PR is open already) so that we can tag the exact version that was used in SV-COMP and then proceed with this.

File: src/2ls/graphml_witness_ext.cpp
Comment (viktormalik) [reply src/2ls/graphml_witness_ext.cpp]: Yeah, that's true. Let's make it a bit simpler, then:
```c
dynamic_cfgt cfg;
if(summary_checker.summary_db.exists(function_name))
{
  const summaryt &summary=summary_checker.summary_db.get(function_name);
  cfg(ssa_unwinder, ssa, summary);
}
else
  cfg(ssa_unwinder, ssa, summaryt());
```

File: src/solver/summary.h
Comment (viktormalik) [inline src/solver/summary.h:59]: Let's make the comment less specific to GOTO instrumentation:
```
// The following fields represent the used abstract domains (containing 
// the template) and the corresponding computed abstract values.
// These can be useful to retrieve information that cannot be easily parsed 
// from the invariant expression.
```

**Review by viktormalik — CHANGES_REQUESTED** [review]

File: src/solver/summary.h
Comment (FrNecas) [reply src/solver/summary.h]: Changed. I recently realised that storing only a single pointer to domain may be a problem. There is only one summary for each function and when adding a summary to `summary_db`, if a summary for the given function already exists, the summaries are joint. Merging two different domains and values probably won't be possible, will it? So my line of thought is that we may need to store a vector of domains/values

Comment (viktormalik) [reply src/solver/summary.h]: I'd stick with one summary per function. If one needs multiple summaries in different domains for a single function, where the result summary is a conjunction of the individual summaries, it's possible to use `product_domain`. If a different composition of summaries should be done, I think that a new combination domain should be created.

**Review by viktormalik — APPROVED** [review]

**Review by peterschrammel — APPROVED** [review]

File: src/domains/ssa_analyzer.h
Comment (peterschrammel) [inline src/domains/ssa_analyzer.h:39]: `inline` is redundant (all functions defined in a header file are `inline`).

**Pull request merged**
